### PR TITLE
feat(filter): add tool_parse filter for Responses API tool classification

### LINKED ISSUE #42: tool_parse: parse and register function/built-in tools
## Parent

Part of praxis-proxy/ai#23

## Scope

- Filter skeleton: module, config, registry registration, `on_request`
- Parse `tools` from request: function tools, `web_search`, `file_search`
- ~Synthesize built-in tool definitions for `web_search`, `file_search`~ moved here https://github.com/praxis-proxy/ai/issues/296
- `tool_choice` — read to understand dispatch behavior, forward as-is to inference server
- Response-side `tools` array: only include function tool definitions (no built-in types leaking into public schema)
- Treat hallucinated/unknown tool names in responses as client-side function calls (not errors)
- Write `responses.tools`, `responses.tool_choice`
- Unit + integration tests

## Why first

Pure request parsing, no external dependencies. Client-side function tools and built-in tools work without any MCP infrastructure.

## Reference

- [Proposal: Responses API Filters](docs/proposals/00354_responses-api-filters.md) — Filter 4

### PR DESCRIPTION
**Pull request by leseb**
## Summary

Add a new `tool_parse` filter that parses the `tools` array and `tool_choice` from Responses API `POST /v1/responses` requests. It classifies each tool by type (function, web_search, file_search, code_interpreter, computer_use, image_generation, tool_search, MCP) and promotes summary facts (`has_tools`, `has_web_search`, `tool_choice`, `function_count`, etc.) to metadata and filter results for downstream `branch_chains` routing decisions.

- Gated to Responses API create path only — non-Responses traffic passes through untouched
- Read-only body access via `StreamBuffer` — does not mutate the request body
- Nameless function tools are counted by discriminator for `has_tools` routing but excluded from `function_tools` metadata
- Unsafe values (oversized, control characters) in `tool_choice` are silently dropped rather than failing the filter
- 82 unit tests + 18 integration tests + 3 example config tests

Closes #42

## Test plan

- [x] Unit tests covering all tool types, tool_choice variants, edge cases, and safety guards
- [x] Integration tests for branch_chains routing on filter results (has_tools, has_web_search, tool_choice)
- [x] Example config test for tool-routing.yaml single-backend pattern
- [x] `make lint` (clippy + nightly fmt) passes
- [x] `make test` (full workspace) passes

### CODE REVIEW

File: apis/src/openai/responses/tool_parse/mod.rs [hunk 1]
```diff
@@ -0,0 +1,340 @@
+// SPDX-License-Identifier: MIT
```
Comment (aslakknutsen) [inline apis/src/openai/responses/tool_parse/mod.rs:1]: Spec: Synthesize built-in tool definitions for web_search, file_search

parser.rs extracts configs but mod.rs never synthesizes function defs or writes metadata

File: apis/src/openai/responses/tool_parse/mod.rs [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        ctx.set_metadata("tool_parse.has_tools", "true");
+    }
+
+    if parsed.function_count > 0 {
+        ctx.set_metadata("tool_parse.function_count", parsed.function_count.to_string());
+    }
+
+    write_tool_presence_metadata(ctx, parsed);
+
+    if let Some(tc) = &parsed.tool_choice {
```
Comment (aslakknutsen) [inline apis/src/openai/responses/tool_parse/mod.rs:176]: Spec: Write responses.tools, responses.tool_choice to metadata

mod.rs writes tool_parse.* keys only; no responses.tools in tool_parse

File: apis/src/openai/responses/tool_parse/mod.rs [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    /// [`FilterError`]: praxis_filter::FilterError
+    pub fn from_config(config: &serde_yaml::Value) -> Result<Box<dyn HttpFilter>, FilterError> {
+        let cfg: ToolParseConfig = parse_filter_config("tool_parse", config)?;
+        let validated = build_config(cfg)?;
+        Ok(Box::new(Self { config: validated }))
+    }
+}
+
+#[async_trait]
+impl HttpFilter for ToolParseFilter {
```
Comment (aslakknutsen) [inline apis/src/openai/responses/tool_parse/mod.rs:95]: Spec: Response-side tools array: function defs only; unknown names as client-side calls

No on_response / on_response_body hooks in tool_parse

**Review by aslakknutsen — CHANGES_REQUESTED** [review]

File: apis/src/openai/responses/tool_parse/mod.rs (hunk 1 above)
Comment (leseb) [reply apis/src/openai/responses/tool_parse/mod.rs:1]: Synthesize is only needed what we do convertion to chat completion, sorry i forgot to update the original issue but i registered https://github.com/praxis-proxy/ai/issues/296 while working on this.

Comment (leseb) [reply apis/src/openai/responses/tool_parse/mod.rs:1]: updated #42 to reflect that

File: apis/src/openai/responses/tool_parse/tests.rs [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+    let req = crate::test_utils::make_request(http::Method::POST, "/v1/responses");
+    let req: &'static praxis_filter::Request = Box::leak(Box::new(req));
+    let mut ctx = crate::test_utils::make_filter_context(req);
+    let mut body = if body_str.is_empty() {
+        None
+    } else {
+        Some(Bytes::from(body_str.to_owned()))
+    };
+
+    drop(filter.on_request_body(&mut ctx, &mut body, true).await.unwrap());
```
Comment (aslakknutsen) [inline apis/src/openai/responses/tool_parse/tests.rs]: run_filter helper drops the on_request_body action without checking it. No test verifies Release on has-tools paths or the missing Release on no-tools/skip paths. Peer filters (validate, rehydrate) explicitly assert Release in body-phase tests. This gap allowed the StreamBuffer retention inconsistency to land undetected.

File: apis/src/openai/responses/tool_parse/parser.rs [hunk 5]
```diff
@@ -0,0 +1,1168 @@
...
+    }
+
+    /// Whether any recognised tool entries were found.
+    ///
+    /// Unlike the classifier's coarse `has_tools` (non-empty array
+    /// check), this counts only entries whose `type` discriminator
+    /// resolved to a known or unknown-but-typed variant. Entries
+    /// missing a `type` field are ignored, so the two checks can
+    /// disagree on malformed arrays.
+    pub(crate) fn has_tools(&self) -> bool {
```
Comment (aslakknutsen) [inline apis/src/openai/responses/tool_parse/parser.rs:221]: ParsedTools::has_tools counts only entries with a recognized or unknown type discriminator (parser.rs:221-223). openai_responses_format uses a coarse non-empty tools array check for stateful routing (proposal 00354 lines 160-162). A body like tools:[{}] may be classified stateful upstream while tool_parse promotes no has_tools, causing branch_chains and stateful routing to disagree. Documented in parser comment but no integration test covers the mismatch.

File: apis/src/openai/responses/tool_parse/mod.rs [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+}
+
+/// Re-promote filter results from metadata written during the body phase.
+///
+/// The pipeline clears `filter_results` after evaluating each filter's
+/// branch conditions. When `tool_parse` is not the first filter in the
+/// chain, a preceding filter's (possibly empty) branch evaluation
+/// clears results before `tool_parse`'s branches are checked. Metadata
+/// persists across phases, so we rebuild `filter_results` from it.
+fn restore_filter_results(ctx: &mut HttpFilterContext<'_>) -> Result<(), FilterError> {
```
Comment (aslakknutsen) [inline apis/src/openai/responses/tool_parse/mod.rs:211]: on_request calls restore_filter_results (mod.rs:110-116) to rebuild filter_results from metadata when a preceding filter clears them. Integration test branch_on_has_web_search_after_preceding_filter indirectly covers the happy path, but there is no unit test for restore_presence_flags, restore_function_count, or restore_tool_choice in isolation. A regression in restore logic would only surface in chained integration scenarios.

**Review by aslakknutsen — CHANGES_REQUESTED** [review]

File: apis/src/openai/responses/tool_parse/mod.rs (hunk 2 above)
Comment (leseb) [reply apis/src/openai/responses/tool_parse/mod.rs:176]: incorrect spec (again) - responses.* is loop state for #26. tool_parse just writes keys for routing - we just classify here

File: apis/src/openai/responses/tool_parse/mod.rs (hunk 3 above)
Comment (leseb) [reply apis/src/openai/responses/tool_parse/mod.rs:95]: incorrect spec (again) - this is #26 - tool_parse just classifies, no hooks, no actions

File: apis/src/openai/responses/tool_parse/parser.rs (hunk 5 above)
Comment (leseb) [reply apis/src/openai/responses/tool_parse/parser.rs:221]: in practice this has 0 side effect, openai_responses_format is right to classify stateful because it doesn't do any validation, tool_parse is correct to decalre tools:[{}] as no-tool. i don't think this is worth a test tbh. the "disagreement" betwen the 2 filters is inconsequential.

File: apis/src/openai/responses/tool_parse/mod.rs (hunk 6 above)
Comment (leseb) [reply apis/src/openai/responses/tool_parse/mod.rs:211]: ok i like isolating the tests failures, so i'll add one test per restore function, covering the happy path

Comment (leseb) [reply apis/src/openai/responses/tool_parse/mod.rs:211]: https://github.com/praxis-proxy/ai/pull/309/commits/266b6e3f00e1719b67c6e1ce26d6718a19c59d98

File: apis/src/openai/responses/tool_parse/tests.rs (hunk 4 above)
Comment (leseb) [reply apis/src/openai/responses/tool_parse/tests.rs]: https://github.com/praxis-proxy/ai/pull/309/commits/266b6e3f00e1719b67c6e1ce26d6718a19c59d98

**Review by aslakknutsen — APPROVED** [review]
